src/mgmt_conn.c

```diff
@@ -717,6 +717,181 @@ internal_snapshot(char *snap)
 	return (B_FALSE);
 }
 
+/*
+ * This fn returns map of snapshots by walking through datasets in json
+ * format. Returns 0 on success.
+ * This will return EAGAIN if the replica is NOT connected to target.
+ */
+static struct json_object *
+uzfs_zvol_fetch_snapshot_jmap(zvol_info_t *zinfo, int *err)
+{
+	char *snapname;
+	boolean_t case_conflict;
+	uint64_t id, pos = 0;
+	int error = 0;
+	zvol_state_t *zv = (zvol_state_t *)zinfo->main_zv;
+	objset_t *os = zv->zv_objset;
+	dsl_dataset_t *ds;
+	dsl_pool_t *dp;
+	struct json_object *jmap = NULL;
+
+	snapname = kmem_zalloc(ZFS_MAX_DATASET_NAME_LEN, KM_SLEEP);
+
+	if ((os == NULL) || (os->os_dsl_dataset == NULL) ||
+	    (os->os_dsl_dataset->ds_dir == NULL) ||
+	    (os->os_dsl_dataset->ds_dir->dd_pool == NULL)) {
+		error = EAGAIN;
+		goto out;
+	}
+
+	jmap = json_object_new_object();
+	dp = os->os_dsl_dataset->ds_dir->dd_pool;
+
+	while (error == 0) {
+		dsl_pool_config_enter(dmu_objset_pool(os), FTAG);
+		error = dmu_snapshot_list_next(os,
+		    ZFS_MAX_DATASET_NAME_LEN, snapname, &id, &pos,
+		    &case_conflict);
+		if (error) {
+			dsl_pool_config_exit(dmu_objset_pool(os), FTAG);
+			break;
+		}
+
+		if (internal_snapshot(snapname)) {
+			dsl_pool_config_exit(dmu_objset_pool(os), FTAG);
+			continue;
+		}
+
+		dsl_pool_config_exit(dmu_objset_pool(os), FTAG);
+
+		json_object_object_add(jmap, snapname, NULL);
+	}
+
+	if ((error = ENOENT) || (error == 0)) {
+		error = 0;
+	} else {
+		json_object_put(jmap);
+		jmap = NULL;
+	}
+out:
+	kmem_free(snapname, ZFS_MAX_DATASET_NAME_LEN);
+	*err = error;
+
+	return (jmap);
+}
+
+/*
+ * This fn returns reference to zinfo's map of snap json_object.
+ * If its NULL, it creates one and returns it
+ * Its the responsibility of caller to release the reference.
+ */
+struct json_object *
+uzfs_zvol_get_snap_jmap_ref(zvol_info_t *zinfo, int *err)
+{
+	boolean_t freeit = B_FALSE;
+	struct json_object *jmap;
+	int error;
+
+	*err = 0;
+	if (zinfo->snap_map == NULL) {
+		jmap = uzfs_zvol_fetch_snapshot_jmap(zinfo, &error);
+
+		if (error) {
+			*err = error;
+			return (NULL);
+		}
+
+		pthread_mutex_lock(&zinfo->snap_map_mutex);
+		if (zinfo->snap_map == NULL)
```

Comment:
We are already having a check right why we need it again

Reply:
this code got changed now